### PR TITLE
Lp 157 support csrf

### PR DESCRIPTION
**Pull request by mattch1**
### JIRA link
https://companieshouse.atlassian.net/browse/LP-157

### Change description

Introduced csrf to the app and to errors handling and updated the error template.

### Work checklist

- [x] Tests added where applicable
- [ ] UI changes look good on mobile
- [ ] UI changes meet accessibility criteria

### Merge instructions

We are committed to keeping commit history clean, consistent and linear. To achieve this commit should be structured as follows:

```
<type>[optional scope]: <description>
```

and contain the following structural elements:

- fix: a commit that patches a bug in your codebase (this correlates with PATCH in semantic versioning),
- feat: a commit that introduces a new feature to the codebase (this correlates with MINOR in semantic versioning),
- BREAKING CHANGE: a commit that has a footer `BREAKING CHANGE:` introduces a breaking API change (correlating with MAJOR in semantic versioning). A BREAKING CHANGE can be part of commits of any type,
- types other than `fix:` and `feat:` are allowed, for example `build:`, `chore:`, `ci:`, `docs:`, `style:`, `refactor:`, `perf:`, `test:`, and others,
- footers other than `BREAKING CHANGE: <description>` may be provided.

### CODE REVIEW

**Comment by mattch1** [comment]
When running the tests I see this:

A worker process has failed to exit gracefully and has been force exited. This is likely caused by tests leaking due to improper teardown. Try running with --detectOpenHandles to find leaks. Active timers can also cause this, ensure that .unref() was called on them.

Not sure what is causing this, cannot find anything that could be blocking this.

**Comment by ch-code-analysis** [comment]
CI: No security warnings found

**Comment by markpit** [comment]
> When running the tests I see this:
> 
> A worker process has failed to exit gracefully and has been force exited. This is likely caused by tests leaking due to improper teardown. Try running with --detectOpenHandles to find leaks. Active timers can also cause this, ensure that .unref() was called on them.
> 
> Not sure what is causing this, cannot find anything that could be blocking this.

adding jest.mock("ioredis"); should hopefully fix it.

File: src/app.ts [hunk 1]
```diff
@@ -5,6 +5,9 @@ import * as path from "path";
 import * as config from "./config";
 import logger from "./utils/logger";
 import router from "./routes";
+import { CsrfProtectionMiddleware } from "@companieshouse/web-security-node";
```
Comment (lduranteau) [inline src/app.ts:8]: it is just a small detail, but is it possible to keep the library imports at the top and then the local imports?

File: src/middlewares/error.middleware.ts [hunk 2]
```diff
@@ -1,13 +1,35 @@
 import { NextFunction, Request, Response } from "express";
 import logger from "../utils/logger";
 import * as config from "../config";
+import { CsrfError } from '@companieshouse/web-security-node';
```
Comment (lduranteau) [inline src/middlewares/error.middleware.ts:4]: same as above for the imports

File: src/test/error-pages.test.ts [hunk 3]
```diff
@@ -4,11 +4,21 @@ import app from "../app";
 import * as config from "../config";
 
 import { get } from "../controllers/start.controller";
+import { CsrfError } from "@companieshouse/web-security-node";
```
Comment (lduranteau) [inline src/test/error-pages.test.ts:7]: same as above for the imports

**Comment by ch-code-analysis** [comment]
CI: No security warnings found

File: src/app.ts (hunk 1 above)
Comment (mattch1) [reply src/app.ts]: Ordered imports

File: src/middlewares/error.middleware.ts (hunk 2 above)
Comment (mattch1) [reply src/middlewares/error.middleware.ts]: Ordered imports

File: src/test/error-pages.test.ts (hunk 3 above)
Comment (mattch1) [reply src/test/error-pages.test.ts]: Ordered imports

**Comment by mattch1** [comment]
> > When running the tests I see this:
> > A worker process has failed to exit gracefully and has been force exited. This is likely caused by tests leaking due to improper teardown. Try running with --detectOpenHandles to find leaks. Active timers can also cause this, ensure that .unref() was called on them.
> > Not sure what is causing this, cannot find anything that could be blocking this.
> 
> adding jest.mock("ioredis"); should hopefully fix it.

Have added ioredis but still see the same message; that might be to do with order in which the mocks and imports appear in the test, but regardless of order this message appears in all cases.

File: src/middlewares/error.middleware.ts [hunk 4]
```diff
@@ -8,6 +10,27 @@ const pageNotFound = (req: Request, res: Response) => {
...
 
+/**
+ * This handler catches any CSRF errors thrown within the application.
+ * If it is not a CSRF, the error is passed to the next error handler.
+ * If it is a CSRF error, it responds with a 403 forbidden status and renders the CSRF error.
+ */
+
+const csrfErrorHandler = (err: CsrfError | Error, req: Request, res: Response, next: NextFunction) => {
+
+  const isRegistration: boolean = req.path.startsWith(config.START_URL);
```
Comment (markpit) [inline src/middlewares/error.middleware.ts:21]: isRegistration is for overseas-entities registration journey

File: src/app.ts [hunk 5]
```diff
@@ -1,6 +1,9 @@
 import express from "express";
 import * as nunjucks from "nunjucks";
 import * as path from "path";
+import Redis from 'ioredis';
```
Comment (markpit) [inline src/app.ts:4]: 'ioredis' has inconsistent quotation marks

**Comment by ch-code-analysis** [comment]
CI: No security warnings found

File: src/middlewares/error.middleware.ts (hunk 4 above)
Comment (mattch1) [reply src/middlewares/error.middleware.ts]: Removed specific code.

File: src/app.ts (hunk 5 above)
Comment (mattch1) [reply src/app.ts]: Using double quotes

**Comment by ch-code-analysis** [comment]
CI: No security warnings found

File: src/middlewares/error.middleware.ts [hunk 6]
```diff
@@ -8,6 +10,25 @@ const pageNotFound = (req: Request, res: Response) => {
   });
 };
 
+/**
+ * This handler catches any CSRF errors thrown within the application.
+ * If it is not a CSRF, the error is passed to the next error handler.
+ * If it is a CSRF error, it responds with a 403 forbidden status and renders the CSRF error.
+ */
+
```
Comment (lduranteau) [inline src/middlewares/error.middleware.ts:18]: extra space between comment and code block

File: src/test/error-pages.test.ts [hunk 7]
```diff
@@ -1,14 +1,26 @@
...
-import * as config from "../config";
-
 import { get } from "../controllers/start.controller";
+import * as config from "../config";
 
 jest.mock("../controllers/start.controller");
+
 const mockGet = get as jest.Mock;
 
+const CSRF_TOKEN_ERROR = "CSRF token mismatch";
```
Comment (lduranteau) [inline src/test/error-pages.test.ts:14]: if a value is only used once, there is no point in creating a constant, it is better to use it directly

**Comment by ch-code-analysis** [comment]
CI: No security warnings found

File: src/test/error-pages.test.ts (hunk 7 above)
Comment (mattch1) [reply src/test/error-pages.test.ts:12]: The constants help to make the test more readable and make it consistent with the layout of other tests, so unless there's a real functional need to change this, I would say we should keep it as it is.

Comment (lduranteau) [reply src/test/error-pages.test.ts:12]: I can't see what makes the test more readable.
A value can be stored in a variable if it is used in at least two different places.

Comment (mattch1) [reply src/test/error-pages.test.ts]: As this seems to be a problem I have changed it to use the values inline.

**Comment by ch-code-analysis** [comment]
CI: No security warnings found

**Comment by mattch1** [comment]
@lduranteau @markpit any updates here?

**Comment by markpit** [comment]
lgtm

**Review by markpit — APPROVED** [review]

File: src/test/error-pages.test.ts (hunk 7 above)
Comment (mattch1) [reply src/test/error-pages.test.ts]: @lduranteau did you want to add anything further if not can we resolve this?

**Review by lduranteau — APPROVED** [review]

**Pull request merged**